fs_attachment/models/ir_attachment.py

```diff
@@ -96,6 +79,46 @@ class IrAttachment(models.Model):
         ondelete="restrict",
     )
 
+    def init(self):
+        res = super().init()
+        # This partial index is used by the register hook to find attachments
+        # to migrate from the odoo filestore to a fs storage
+        query = """
+            CREATE INDEX IF NOT EXISTS
+            ir_attachment_no_fs_storage_code
+            ON ir_attachment (fs_storage_code)
+            WHERE fs_storage_code IS NULL;
+        """
+        self.env.cr.execute(query)
+        return res
+
+    def _register_hook(self):
+        super()._register_hook()
+        fs_storage_codes = self._get_storage_codes()
+        # ignore if we are not using an object storages
+        if not fs_storage_codes:
+            return
+        curframe = inspect.currentframe()
+        calframe = inspect.getouterframes(curframe, 2)
+        # the caller of _register_hook is 'load_modules' in
+        # odoo/modules/loading.py
+        load_modules_frame = calframe[1][0]
+        # 'update_module' is an argument that 'load_modules' receives with a
+        # True-ish value meaning that an install or upgrade of addon has been
+        # done during the initialization. We need to move the attachments that
+        # could have been created or updated in other addons before this addon
+        # was loaded
+        update_module = load_modules_frame.f_locals.get("update_module")
```

Comment:
Is it not too much magic? I don't know, I think I prefer to do these kind of operations that can potentially lose data explicitly.

Reply:
Couldn't this be simplified by simply running it every time?
If there's nothing to do the cost shouldn't be too expensive.

The manual solution has the issue to be an additional hurdle for deployment that requires to be well documented, or the user might face some troubles that might be bad for adoption.